pgjdbc/src/main/java/org/postgresql/jdbc/TimestampUtils.java

```diff
@@ -1294,7 +1297,14 @@ public Time convertToTime(long millis, TimeZone tz) {
     cal.set(Calendar.ERA, GregorianCalendar.AD);
     cal.set(Calendar.YEAR, 1970);
     cal.set(Calendar.MONTH, 0);
-    cal.set(Calendar.DAY_OF_MONTH, 1);
+    // We allow 24:00:00, however all the other dates go with 1970-01-01
```

Comment:
As I said before, this is invalid. `java.sql.Time` says it must be _normalized_.  So, PostgreSQL is handing you something unrepresentable in `java.sql.Time`.

Reply:
That might be true but we can't just ignore it because 'java.sql.time' doesn't deal with it

Reply:
This code is converting a PostgreSQL `time` to a `java.sql.Time` wrapping it to zero is not ignoring it. It's doing the conversion correctly.

Reply:
Suppose we have `23:59:59.999999999`.
I guess PostgreSQL can handle something like `23:59:59.99999` (or whatever digits it can).

`java.sql.Time` can have `23:59:59.999` only while `java.time.LocalTime` can represent `23:59:59.999999999`.

I would say it would be quite surprising to return `00:00:00` `java.sql.Time` and `23:59:59.999999999` `java.time.LocalTime` at the same time.

In the same sense, I would use "round to the closest" approach (i.e. `24:00:00` -> `23:59:59.999999999`) rather than `24:00:00` -> `00:00:00`

Reply:
I'd say returning `00:00:00` for `java.sql.Time` and `23:59:59.99999999` for `LocalTime` is exactly what I'd expect.  After all that's a precision argument.  Adding `.000000001` second to the `LocalTime` would then equal `00:00:00`.  Also, rounding the `LocalTime` time to millisecond precision would yield `00:00:00` as well.

24 hour time wraps. As I stated in the issue thread `24:00:00 == 00:00:00` is a _true_ equality.  `24:00:00 == 23:59:59.999` is not true.

Reply:
`1=0.9999(9)` is a true equality by the way.

Reply:
Lol. That equality only holds true when 9 the repeats infinitely. Sadly our computers are not there yet.

Reply:
ISO8601 allows 24:00:00.
Relevant issue in Ruby on Rails: https://github.com/rails/rails/issues/7125.

Relevant commit to the OpenJDK: http://hg.openjdk.java.net/threeten/threeten/jdk/rev/5c3984ba10a3
I have not analyzed it yet, however it does seem to target 24:00:00 case.
It looks like "24:00:00 is parsed as hour=0, minute=0, and `DateTimeFormatter#parsedExcessDays` signals the extra day".

Reply:
`DateTimeFormatter` is a pretty complete parser/formatter, I know because I recently rewrote all of `pgjdbc-ng`'s time/date/timestamp handling with it.  It has lenient modes and strict modes and scores of other options for parsing text.

When it converts the object to and jsr310 object the fields are required to be exact (that's what ResolverStyle helps with).  Point is, just because the formatter allows parsing it and provides you information doesn't mean it can produce a valid jsr310 object. 

`LocalTime`, in this case, ignores such a field and uses the _wrapped_ time value.

Reply:
From ISO8601

> [hh] refers to a zero-padded hour between 00 and 24 (where 24 is only used to denote midnight at the end of a calendar day)

This might shed some light on why 24 is allowed in this format (and thus PostgreSQL). Unfortunately, none of this changes the fact that `java.sql.Time` doesn't allow an "end of day" value.

Reply:
the litmus test for me would be. If I read 24:00:00 in from the database, can I save it back? If not then I'd say we throw an exception if someone attempts to read 24:00:00 into  ```java.sql.Time```

Reply:
It's a bit more murky an answer than that... look at #1389.

Although the value can be stored in `java.sql.Time` it is not allowed.  Then comparisons on what _seem_ to be equal values result in an inequality.  This actually has much wider reaching ramifications than just this one test.

FYI, I'll post a WIP PR for #1389 that provides an updated `TimezoneTest` that properly creates `Date` and `Time` values. The fix seems to be complicated because it touches all three `time`, `date` and `timestamp` encoding/decoding so I didn't attempt it.